python/pyspark/sql/group.py

```diff
@@ -243,14 +244,25 @@ def apply(self, udf):
         func = udf.func
         returnType = udf.returnType
 
-        # The python executors expects the function to take a list of pd.Series as input
+        # The python executors expects the function to use pd.Series as input and output
         # So we to create a wrapper function that turns that to a pd.DataFrame before passing
-        # down to the user function
+        # down to the user function, then turn the result pd.DataFrame back into pd.Series
         columns = df.columns
 
         def wrapped(*cols):
             import pandas as pd
-            return func(pd.concat(cols, axis=1, keys=columns))
+            result = func(pd.concat(cols, axis=1, keys=columns))
+            if not isinstance(result, pd.DataFrame):
+                raise TypeError("Return type of the user-defined function should be "
+                                "Pandas.DataFrame, but is {}".format(type(result)))
+            if not len(result.columns) == len(returnType):
+                raise RuntimeError(
+                    "Number of columns of the returned Pandas.DataFrame "
+                    "doesn't match specified schema. "
+                    "Expected: {} Actual: {}".format(len(returnType), len(result.columns)))
+            arrow_return_types = (to_arrow_type(field.dataType) for field in returnType)
+            return [(result[result.columns[i]], arrow_type)
+                    for i, arrow_type in enumerate(arrow_return_types)]
```

Comment:
I made a few changes here and `arrow_return_types` is a generator, so that columns only had to be iterated over once